### ISSUE TITLE
KoinAppAlreadyStartedException after closing the app and relaunching it quickly

### ISSUE THREAD

**Issue by armond-avanes** [InsertKoinIO/koin#1864]
**Describe the bug**
I'm using Koin in my Compose Multiplatform project and trying to keep as much as possible in `commonMain`, so I'm using `KoinApplication(...)` in my top layer composable function.

When I close the app (Android) and relaunch it quickly, I get `KoinAppAlreadyStartedException`. I believe that when I close the app, the activity is destroyed but the app itself is still alive for a couple of more seconds... in that case, when I relaunch it quickly, the app itself is not reinitiated (already exists) but the activity is recreated (because it has been destroyed). When `KoinApplication(...)` is called and it detects that the Koin has already been started in the past, it throws the exception.

I know I can make use of `KoinContext()` and instead start the Koin in "Application.onCreate()" to fix this issue. But then I will need to repeat a similar logic for iOS and Desktop which is what I'm trying to avoid.

**To Reproduce**
1. Close the app
2. Quickly relaunch it
3. See error

**Expected behavior**
`KoinApplication` should handle this situation gracefully.

**Koin module and version:**
- `koin-core:3.6.0-Beta2`
- `koin-compose:1.2.0-Beta2`

**Comment by Nek-12** [comment]
By starting Koin in composition you absolve yourself of any chance of injecting dependencies outside composition. The correct approach is to start Koin in the application and make a function in shared module that starts koin and call it from swift.

**Comment by armond-avanes** [comment]
@Nek-12 I understand your point, but `KoinApplication()` is a valid Koin [composable] API and it's not just a hacky way of doing things. If there is an API for something, it should work. Otherwise the API can be marked as deprecated, and the documentation should be updated to explain that there is no platform independent way to start Koin in Compose.

**Comment by SebastianFelon** [comment]
I totally agree with @armond-avanes on this. I've just wasted a lot of time trying to find a fix for keeping Koin initialization in multiplatform common only as it seemed a natural continuation for platform independence.

**Comment by harshal-msf** [comment]
Totally aggree with @armond-avanes But there is one workaround to properly initializa koin dependencies like below. this will reduce your effort to add platform specific code in android and ios

`KoinContext(context = koinApplication {
        modules(sharedKoinModule)
    }.koin) {
        AppContent()
    }`

**Comment by amalhanaja** [comment]
I have similar problems when using `KoinApplication()`. 

Why `KoinApplication()` is throw an exception instead of useing the existing instances when it's already exists?.

Here's my workaround

```kotlin
@Composable
private fun KoinApp(
    application: KoinAppDeclaration,
    content: @Composable () -> Unit,
) {
    val koin = remember(application) {
        val alreadyExists = KoinPlatformTools.defaultContext().getOrNull() != null
        if (alreadyExists) {
            KoinPlatform.getKoin()
        } else {
            startKoin(application).koin
        }
    }
    KoinContext(context = koin, content = content)
}
```

**Comment by lucasmacielc** [comment]
Enabling the `Don't keep activities` option in the developer menu makes reproducing this issue much easier.

Btw, I'm using the same approach as @amalhanaja, as I see no reason not to do so for now.

**Comment by arnaudgiuliani** [comment]
It shouldn't be the same in 4.0. It has been updated.
